<a href="https://colab.research.google.com/github/animesh-11/AI_ML/blob/main/Linear_and_Logistic_Regression_Regularisation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Load the breast cancer dataset, split it into training and testing sets, scale the features, train a Logistic Regression model, and evaluate its performance. Additionally, load the diabetes dataset, split it into training and testing sets, scale the features, train Linear, Lasso (with alpha=1), and Ridge (with alpha=1) regression models, and evaluate their performance using Mean Squared Error and model coefficients.

## Load Breast Cancer Dataset

### Subtask:
Load the breast cancer dataset from `sklearn.datasets.load_breast_cancer()`.


**Reasoning**:
Load the breast cancer dataset from scikit-learn, extract features and target into pandas DataFrame and Series respectively.



In [1]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

# 2. Call load_breast_cancer() and store the returned object
cancer_data = load_breast_cancer()

# 3. Extract the feature data (X) and store it in a pandas DataFrame
X = pd.DataFrame(cancer_data.data, columns=cancer_data.feature_names)

# 4. Extract the target data (y) and store it in a pandas Series
y = pd.Series(cancer_data.target, name='target')

print("Features (X) head:")
print(X.head())
print("\nTarget (y) head:")
print(y.head())

Features (X) head:
   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perime

## Split Breast Cancer Data

### Subtask:
Split the breast cancer dataset into 80:20 training and testing sets using `train_test_split` with `random_state=9001`.


**Reasoning**:
The subtask requires splitting the dataset into training and testing sets. This can be achieved using the `train_test_split` function from `sklearn.model_selection`. The `X` and `y` variables from the previous step are the features and target respectively, which will be used for splitting. Printing the shapes will confirm the split happened correctly.



In [2]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=9001)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (455, 30)
Shape of X_test: (114, 30)
Shape of y_train: (455,)
Shape of y_test: (114,)


## Scale Breast Cancer Data

### Subtask:
Scale the features of the breast cancer dataset using `MinMaxScaler`.


**Reasoning**:
The subtask requires scaling the features of the breast cancer dataset using `MinMaxScaler`. This involves importing the scaler, instantiating it, fitting it to the training data, and then transforming both the training and testing sets. I will print the first few rows of the scaled dataframes to verify the scaling operation.



In [3]:
from sklearn.preprocessing import MinMaxScaler

# Instantiate a MinMaxScaler object
scaler = MinMaxScaler()

# Fit the scaler to the training data
scaler.fit(X_train)

# Transform both the training and testing data
X_train_scaled = pd.DataFrame(scaler.transform(X_train), columns=X_train.columns)
X_test_scaled = pd.DataFrame(scaler.transform(X_test), columns=X_test.columns)

print("X_train_scaled head:")
print(X_train_scaled.head())
print("\nX_test_scaled head:")
print(X_test_scaled.head())

X_train_scaled head:
   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0     0.300961      0.189619        0.295073   0.159703         0.498059   
1     0.514411      0.098616        0.489323   0.359958         0.232464   
2     0.282503      0.195156        0.271923   0.157031         0.432157   
3     0.254579      0.395156        0.248704   0.136840         0.316783   
4     0.210564      0.173356        0.202267   0.108717         0.395053   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0          0.411386        0.274156             0.372045       0.445658   
1          0.131250        0.104245             0.222471       0.030477   
2          0.184191        0.144348             0.179446       0.323749   
3          0.230262        0.141065             0.199414       0.438758   
4          0.151862        0.082153             0.153088       0.326049   

   mean fractal dimension  ...  worst radius  worst texture  worst peri

## Train Logistic Regression Model

### Subtask:
Train a `LogisticRegression` model (using the default solver) on the scaled breast cancer training data.


**Reasoning**:
The subtask requires training a Logistic Regression model. This involves importing the `LogisticRegression` class, instantiating it, and then fitting it to the scaled training features (`X_train_scaled`) and target (`y_train`).



In [4]:
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression model object
model = LogisticRegression()

# Fit the LogisticRegression model to the scaled training data
model.fit(X_train_scaled, y_train)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


## Evaluate Logistic Regression Model

### Subtask:
Evaluate the performance of the trained `LogisticRegression` model on the scaled breast cancer test data using accuracy, precision, recall, and F1-score.

**Reasoning**:
The subtask is to evaluate the performance of the trained Logistic Regression model. This involves importing necessary metrics, making predictions on the scaled test data, and then calculating accuracy, precision, recall, and F1-score.



In [5]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the scaled test data
y_pred = model.predict(X_test_scaled)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Accuracy: 1.0000
Precision: 1.0000
Recall: 1.0000
F1-Score: 1.0000


## Load Diabetes Dataset

### Subtask:
Load the diabetes dataset from `sklearn.datasets.load_diabetes()`.

## Load Diabetes Dataset

### Subtask:
Load the diabetes dataset from `sklearn.datasets.load_diabetes()`.

**Reasoning**:
The subtask requires loading the diabetes dataset. This involves importing `load_diabetes` from `sklearn.datasets`, calling the function, and then storing the features and target into pandas DataFrames/Series.



In [6]:
import pandas as pd
from sklearn.datasets import load_diabetes

# Call load_diabetes() and store the returned object
diabetes_data = load_diabetes()

# Extract the feature data (X) and store it in a pandas DataFrame
X_diabetes = pd.DataFrame(diabetes_data.data, columns=diabetes_data.feature_names)

# Extract the target data (y) and store it in a pandas Series
y_diabetes = pd.Series(diabetes_data.target, name='target')

print("Diabetes Features (X_diabetes) head:")
print(X_diabetes.head())
print("\nDiabetes Target (y_diabetes) head:")
print(y_diabetes.head())

Diabetes Features (X_diabetes) head:
        age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  
0 -0.002592  0.019907 -0.017646  
1 -0.039493 -0.068332 -0.092204  
2 -0.002592  0.002861 -0.025930  
3  0.034309  0.022688 -0.009362  
4 -0.002592 -0.031988 -0.046641  

Diabetes Target (y_diabetes) head:
0    151.0
1     75.0
2    141.0
3    206.0
4    135.0
Name: target, dtype: float64


## Split Diabetes Data

### Subtask:
Split the diabetes dataset into 80:20 training and testing sets using `train_test_split` with `random_state=42`.

**Reasoning**:
The subtask requires splitting the diabetes dataset into training and testing sets. This can be achieved using the `train_test_split` function from `sklearn.model_selection` on the `X_diabetes` and `y_diabetes` variables. Printing the shapes will confirm the split happened correctly.



In [7]:
from sklearn.model_selection import train_test_split

# Split the diabetes dataset into training and testing sets
X_diabetes_train, X_diabetes_test, y_diabetes_train, y_diabetes_test = train_test_split(X_diabetes, y_diabetes, test_size=0.2, random_state=42)

print("Shape of X_diabetes_train:", X_diabetes_train.shape)
print("Shape of X_diabetes_test:", X_diabetes_test.shape)
print("Shape of y_diabetes_train:", y_diabetes_train.shape)
print("Shape of y_diabetes_test:", y_diabetes_test.shape)

Shape of X_diabetes_train: (353, 10)
Shape of X_diabetes_test: (89, 10)
Shape of y_diabetes_train: (353,)
Shape of y_diabetes_test: (89,)


## Scale Diabetes Data

### Subtask:
Scale the features of the diabetes dataset using `MinMaxScaler`.

**Reasoning**:
The subtask requires scaling the features of the diabetes dataset using `MinMaxScaler`. This involves importing the scaler, instantiating it, fitting it to the training data, and then transforming both the training and testing sets. I will print the first few rows of the scaled dataframes to verify the scaling operation.



In [8]:
from sklearn.preprocessing import MinMaxScaler

# Instantiate a MinMaxScaler object
scaler_diabetes = MinMaxScaler()

# Fit the scaler to the training data
scaler_diabetes.fit(X_diabetes_train)

# Transform both the training and testing data
X_diabetes_train_scaled = pd.DataFrame(scaler_diabetes.transform(X_diabetes_train), columns=X_diabetes_train.columns)
X_diabetes_test_scaled = pd.DataFrame(scaler_diabetes.transform(X_diabetes_test), columns=X_diabetes_test.columns)

print("X_diabetes_train_scaled head:")
print(X_diabetes_train_scaled.head())
print("\nX_diabetes_test_scaled head:")
print(X_diabetes_test_scaled.head())

X_diabetes_train_scaled head:
        age  sex       bmi        bp        s1        s2        s3        s4  \
0  0.816667  1.0  0.405172  0.690141  0.544503  0.524900  0.220779  0.423131   
1  0.450000  1.0  0.284483  0.323944  0.335079  0.406375  0.103896  0.564175   
2  0.516667  0.0  0.556034  0.859155  0.356021  0.315737  0.337662  0.282087   
3  0.366667  0.0  0.215517  0.338028  0.198953  0.181275  0.467532  0.141044   
4  0.383333  0.0  0.094828  0.126761  0.267016  0.197211  0.571429  0.000000   

         s5        s6  
0  0.590930  0.500000  
1  0.486609  0.575758  
2  0.551160  0.454545  
3  0.293517  0.030303  
4  0.342588  0.348485  

X_diabetes_test_scaled head:
        age  sex       bmi        bp        s1        s2        s3        s4  \
0  0.700000  0.0  0.331897  0.394366  0.890052  0.765936  0.428571  0.423131   
1  0.916667  0.0  0.504310  0.549296  0.319372  0.314741  0.363636  0.141044   
2  0.783333  1.0  0.340517  0.408451  0.806283  0.522908  0.558442  0.28208

## Train Regression Models for Diabetes Dataset

### Subtask:
Train Linear, Lasso (with `alpha=1`), and Ridge (with `alpha=1`) regression models on the scaled diabetes training data.

**Reasoning**:
The subtask requires training Linear, Lasso, and Ridge regression models. This involves importing the respective classes, instantiating each model, setting the `alpha` parameter to 1 for Lasso and Ridge, and fitting them to the scaled training data (`X_diabetes_train_scaled`, `y_diabetes_train`).



In [9]:
from sklearn.linear_model import LinearRegression, Lasso, Ridge

# Instantiate and train Linear Regression model
linear_model = LinearRegression()
linear_model.fit(X_diabetes_train_scaled, y_diabetes_train)
print("Linear Regression model trained successfully.")

# Instantiate and train Lasso Regression model (alpha=1)
lasso_model = Lasso(alpha=1)
lasso_model.fit(X_diabetes_train_scaled, y_diabetes_train)
print("Lasso Regression model trained successfully.")

# Instantiate and train Ridge Regression model (alpha=1)
ridge_model = Ridge(alpha=1)
ridge_model.fit(X_diabetes_train_scaled, y_diabetes_train)
print("Ridge Regression model trained successfully.")

Linear Regression model trained successfully.
Lasso Regression model trained successfully.
Ridge Regression model trained successfully.


## Evaluate Regression Models for Diabetes Dataset

### Subtask:
Evaluate the performance of the trained Linear, Lasso, and Ridge regression models on the scaled diabetes test data using Mean Squared Error and print their coefficients.

**Reasoning**:
The subtask requires evaluating the performance of the trained regression models. This involves making predictions on the scaled test data, calculating the Mean Squared Error for each model, and printing their coefficients.



In [10]:
from sklearn.metrics import mean_squared_error

# Evaluate Linear Regression model
y_pred_linear = linear_model.predict(X_diabetes_test_scaled)
mse_linear = mean_squared_error(y_diabetes_test, y_pred_linear)
print(f"Linear Regression - Mean Squared Error: {mse_linear:.4f}")
print("Linear Regression - Coefficients:", linear_model.coef_)

# Evaluate Lasso Regression model
y_pred_lasso = lasso_model.predict(X_diabetes_test_scaled)
mse_lasso = mean_squared_error(y_diabetes_test, y_pred_lasso)
print(f"\nLasso Regression (alpha=1) - Mean Squared Error: {mse_lasso:.4f}")
print("Lasso Regression (alpha=1) - Coefficients:", lasso_model.coef_)

# Evaluate Ridge Regression model
y_pred_ridge = ridge_model.predict(X_diabetes_test_scaled)
mse_ridge = mean_squared_error(y_diabetes_test, y_pred_ridge)
print(f"\nRidge Regression (alpha=1) - Mean Squared Error: {mse_ridge:.4f}")
print("Ridge Regression (alpha=1) - Coefficients:", ridge_model.coef_)

Linear Regression - Mean Squared Error: 2900.1936
Linear Regression - Coefficients: [   8.2612689   -23.06446772  135.63561351   84.9935688  -244.80178532
  162.87932677   46.32729542   72.03113268  191.18672297   13.3055384 ]

Lasso Regression (alpha=1) - Mean Squared Error: 2809.5035
Lasso Regression (alpha=1) - Coefficients: [  0.         -11.39818989 134.43774897  64.07795621  -0.
  -0.         -51.42795174   0.         100.17927032   0.        ]

Ridge Regression (alpha=1) - Mean Squared Error: 2865.1816
Ridge Regression (alpha=1) - Coefficients: [  9.85156564 -22.34898822 130.29735822  80.82738282 -40.69503261
 -19.51884299 -49.92923696  43.34048984 107.55652116  19.58621215]


## Summary:

### Q&A
1.  **How did the Logistic Regression model perform on the breast cancer dataset?**
    The Logistic Regression model achieved perfect scores on the breast cancer test set: 1.0000 Accuracy, 1.0000 Precision, 1.0000 Recall, and 1.0000 F1-Score.

2.  **How did the Linear, Lasso, and Ridge regression models perform on the diabetes dataset?**
    *   **Linear Regression:** Achieved a Mean Squared Error (MSE) of 2900.1936.
    *   **Lasso Regression (alpha=1):** Achieved an MSE of 2809.5035.
    *   **Ridge Regression (alpha=1):** Achieved an MSE of 2865.1816.

### Data Analysis Key Findings
*   The breast cancer dataset was successfully loaded, split into 80:20 training (455 samples) and testing (114 samples) sets, and its features scaled using `MinMaxScaler`.
*   The Logistic Regression model trained on the scaled breast cancer data demonstrated exceptional performance, achieving perfect scores (1.0000 for Accuracy, Precision, Recall, and F1-Score) on the test set.
*   The diabetes dataset was loaded, split into 80:20 training (353 samples) and testing (89 samples) sets, and its features scaled using `MinMaxScaler`.
*   For the diabetes dataset, Lasso Regression (with `alpha=1`) achieved the lowest Mean Squared Error (MSE = 2809.5035) among the three regression models, outperforming Linear Regression (MSE = 2900.1936) and Ridge Regression (MSE = 2865.1816) on the test set.
*   The coefficients of the Lasso Regression model showed feature selection, with some coefficients becoming zero, indicating that certain features were deemed less important for the prediction.

### Insights or Next Steps
*   The perfect scores of the Logistic Regression model on the breast cancer dataset suggest that the classes are highly separable, or it might indicate potential data leakage or overfitting if not cross-validated. Further investigation with cross-validation or different evaluation strategies could confirm the model's robustness.
*   For the diabetes dataset, Lasso Regression showed a slight improvement in performance (lower MSE) compared to Linear and Ridge Regression, while also performing feature selection. This suggests that regularization techniques are beneficial for this dataset, and further hyperparameter tuning for `alpha` in Lasso and Ridge models could lead to even better performance.
